CMakeLists.txt

```diff
@@ -6,138 +6,555 @@
+      if(EXISTS ${LLVM_MAIN_SRC_DIR}/utils/llvm-lit)
+        add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/llvm-lit utils/llvm-lit)
+      endif()
+
+    if(NOT LLVM_UTILS_PROVIDED)
+      add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/FileCheck utils/FileCheck)
+      add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/count utils/count)
+      add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/not utils/not)
+      set(LLVM_UTILS_PROVIDED ON)
+      set(FLANG_TEST_DEPS FileCheck count not)
+    endif()
+
+    set(UNITTEST_DIR ${LLVM_MAIN_SRC_DIR}/utils/unittest)
+    
+    if (EXISTS ${UNITTEST_DIR}/googletest/include/gtest/gtest.h
+        AND NOT EXISTS 
+          ${LLVM_LIBRARY_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}gtest${CMAKE_STATIC_LIBRARY_SUFFIX}
+        AND EXISTS ${UNITTEST_DIR}/CMakeLists.txt)
+          add_subdirectory(${UNITTEST_DIR} utils/unittest)
+      endif()
+    else()
+      # Search for an installed lit.
+      find_program(LLVM_LIT
+                   NAMES llvm-lit lit.py lit
+                   PATHS "${LLVM_MAIN_SRC_DIR}/utils/lit"
+                   DOC "Path to lit.py")
+    endif()
 
-if(LINK_WITH_FIR)
+    if (LLVM_LIT)
+      # Define the default arguments to use with 'lit', and provide an option for the user
+      # to override the arguments.
+      set(LIT_ARGS_DEFAULT "-sv")
+
+      if (MSVC OR XCODE)
+        set(LIT_ARGS_DEFAULT "${LIT_ARGS_DEFAULT} --no-progress-bar")
+      endif()
+      
+      set(LLVM_LIT_ARGS "${LIT_ARGS_DEFAULT}" CACHE STRING "Default options for lit")
+
+      # On Win32 hosts, provide an option to specify the path to the GnuWin32 tools.
+      if( WIN32 AND NOT CYGWIN )
+        set(LLVM_LIT_TOOLS_DIR "" CACHE PATH "Path to GnuWin32 tools")
+      endif()
+
+    else()
+      set(LLVM_INCLUDE_TESTS OFF)
+    endif()
+
+  endif()
+
+  set(FLANG_BUILT_STANDALONE 1)
+  set(BACKEND_PACKAGE_STRING "LLVM ${LLVM_PACKAGE_VERSION}")
+else()
+  set(BACKEND_PACKAGE_STRING "${PACKAGE_STRING}")
+  set(MLIR_BINARY_DIR ${LLVM_BINARY_DIR}/tools/mlir)
+  set(MLIR_SOURCE_DIR ${LLVM_SOURCE_DIR}/../mlir)
+endif()
+
+# Make sure that the flang source directory is on the current cmake module path so that 
+# we can include flang-centric cmake files from within the project. 
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules")
+
+if(LLVM_ENABLE_LIBXML2)
+  # Don't look for libxml if we're using MSan, since uninstrumented third party
+  # code may call MSan interceptors like strlen, leading to false positives.
+  if(NOT LLVM_USE_SANITIZER MATCHES "Memory.*")
+    set (LIBXML2_FOUND 0)
+    find_package(LibXml2 2.5.3 QUIET)
+    if (LIBXML2_FOUND)
+      set(FLANG_HAVE_LIBXML 1)
+    endif()
+  endif()
+endif()
+
+option(LINK_WITH_FIR "Link driver with FIR and LLVM" ON)
+if (LINK_WITH_FIR)
   message(STATUS "Linking driver with FIR and LLVM")
   llvm_map_components_to_libnames(LLVM_COMMON_LIBS support)
   message(STATUS "LLVM libraries: ${LLVM_COMMON_LIBS}")
 endif()
```

Comment:
The LINK_WITH_FIR option was a temporary measure to get past problems building against the right LLVM. I think it can be eliminated.